pkg/timelock/timelock.go

```diff
@@ -132,63 +132,81 @@ func (tw *Worker) Listen(ctx context.Context) error {
+	go handleOSSignal(stopCh)
 
+	// Update timelock-worker default scheduler delay.
 	tw.updateSchedulerDelay(time.Duration(tw.pollPeriod) * time.Second)
+
+	// Run the scheduler to add/del operations in a thread-safe way.
 	go tw.runScheduler(ctx)
-	tw.startLog()
 
-	// Shutdown gracefully based on OS interrupts.
-	go func() {
-		for {
-			handleOSSignal(<-sigCh, stopCh)
-		}
-	}()
+	// Initialize the subscription.
+	logCh := make(chan types.Log)
+	if err := tw.subscribeAndProcessLogs(ctx, logCh, stopCh); err != nil {
+		tw.logger.Error().Err(err).Msg("failed to subscribe and process logs.")
+		return err
+	}
+
+	defer close(stopCh)
+	defer close(logCh)
+	defer ctx.Done()
+	defer tw.dumpOperationStore(time.Now)
+
+	return nil
+}
 
-	// FilterQuery to be feed to the subscription and FilterLogs.
-	query := ethereum.FilterQuery{
+func (tw *Worker) setupFilterQuery() ethereum.FilterQuery {
+	return ethereum.FilterQuery{
 		Addresses: tw.address,
 		FromBlock: tw.fromBlock,
 	}
+}
 
-	tw.logger.Info().Msgf("Starting subscription")
-	// Create the new subscription with the predefined query.
+// subscribeAndProcessLogs creates a subscription and processes logs based on a filter query.
+func (tw *Worker) subscribeAndProcessLogs(ctx context.Context, logCh chan types.Log, stopCh chan string) error {
+	query := tw.setupFilterQuery()
+
+	// SubscribeFilterLogs creates an asynchronous subscription to the events.
+	// It receives all the new events.
+	tw.logger.Info().Msg("starting subscription")
 	sub, err := tw.ethClient.SubscribeFilterLogs(ctx, query, logCh)
 	if err != nil {
 		return err
 	}
+	defer sub.Unsubscribe()
 
-	// Read events by FilterLogs. This method calls eth_getLogs under the hood.
+	// FilterLogs starts filtering the logs at fromBlock, gathering the historical data.
+	// This is needed to guarantee that all the events are gathered, even in scenarios where the worker crashes.
 	filter, err := tw.ethClient.FilterLogs(ctx, query)
 	if err != nil {
 		return err
 	}
 
+	// Process incoming historical logs in a separate goroutine.
+	wg.Add(1)
 	go func() {
-		for _, l := range filter {
-			logCh <- l
+		for _, log := range filter {
+			logCh <- log
 		}
+		wg.Done()
 	}()
 
-	// Setting readyStatus here because we want to make sure subscription is up.
-	tw.logger.Info().Msgf("Initial subscription complete")
-	SetReadyStatus(HealthStatusOK)
-
 	// This is the goroutine watching over the subscription.
 	// We want wg.Done() to cancel the whole execution, so don't add more than 1 to wg.
 	// Also, when receiving an event that creates an error, skip the event and
 	// continue processing the rest, as an external operator can cancel the faulty event.
-	loop := true
 	wg.Add(1)
 	go func() {
-		for loop {
+	MainLoop:
+		for {
 			select {
 			case log := <-logCh:
 				// Decode the log into an event using the ABI exposed in Timelock.go
-				event, err := tw.ABI.EventByID(log.Topics[0])
+				event, err := tw.abi.EventByID(log.Topics[0])
```

Comment:
I would move all the logic for handling a log into a separate method like `HandleLog` and define a `LogHandler` interface.  That way you can launch a routine for the finite slice returned from `FilterQuery` and use the same logic on it.  Also you can then unit test each log handling routine if you add a `LogHandler` dependency.